src/creation/creator.rs

```diff
@@ -0,0 +1,277 @@
+use crate::{
+    nodes::{NodeCount, NodeIndex, NodeMap},
+    units::{ControlHash, PreUnit, Unit},
+    Hasher, Round,
+};
+use log::trace;
+
+pub(super) struct Creator<H: Hasher> {
+    node_id: NodeIndex,
+    n_members: NodeCount,
+    candidates_by_round: Vec<NodeMap<Option<H::Hash>>>,
+    n_candidates_by_round: Vec<NodeCount>, // len of this - 1 is the highest round number of all known units
+}
+
+impl<H: Hasher> Creator<H> {
+    pub(super) fn new(node_id: NodeIndex, n_members: NodeCount) -> Self {
+        Creator {
+            node_id,
+            n_members,
+            candidates_by_round: vec![NodeMap::new_with_len(n_members)],
+            n_candidates_by_round: vec![NodeCount(0)],
+        }
+    }
+
+    pub(super) fn current_round(&self) -> Round {
+        (self.n_candidates_by_round.len() - 1) as Round
+    }
+
+    // initializes the vectors corresponding to the given round (and all between if not there)
+    fn init_round(&mut self, round: Round) {
+        if round > self.current_round() {
+            let new_size = (round + 1).into();
+            self.candidates_by_round
+                .resize(new_size, NodeMap::new_with_len(self.n_members));
+            self.n_candidates_by_round.resize(new_size, NodeCount(0));
+        }
+    }
+
+    // only returns an error if can_create(round) was false
+    pub(super) fn create_unit(&self, round: Round) -> Result<(PreUnit<H>, Vec<H::Hash>), ()> {
+        if !self.can_create(round) {
+            return Err(());
+        }
+        let parents = {
+            if round == 0 {
+                NodeMap::new_with_len(self.n_members)
+            } else {
+                self.candidates_by_round[(round - 1) as usize].clone()
+            }
+        };
+
+        let control_hash = ControlHash::new(&parents);
+        let parent_hashes: Vec<H::Hash> = parents.into_iter().flatten().collect();
+
+        let new_preunit = PreUnit::new(self.node_id, round, control_hash);
+        trace!(target: "AlephBFT-creator", "Created a new unit {:?} at round {:?}.", new_preunit, round);
+        Ok((new_preunit, parent_hashes))
+    }
+
+    pub(super) fn add_unit(&mut self, unit: &Unit<H>) {
+        let round = unit.round();
+        let pid = unit.creator();
+        let hash = unit.hash();
+        self.init_round(round);
+        if self.candidates_by_round[round as usize][pid].is_none() {
+            // passing the check above means that we do not have any unit for the pair (round, pid) yet
+            self.candidates_by_round[round as usize][pid] = Some(hash);
+            self.n_candidates_by_round[round as usize] += NodeCount(1);
+        }
+    }
+
+    /// To create a new unit, we need to have at least floor(2*N/3) + 1 parents available in previous round.
+    /// Additionally, our unit from previous round must be available.
+    pub(super) fn can_create(&self, round: Round) -> bool {
```

Comment:
Consider making this method private since `create_unit` now returns `Result`. It seems that the only non-trivial usage is by `wait_until_ready`.

Reply:
That is the only usage, but I don't quite see how to remove it without significantly changing the logic, and I don't have an idea for a good version of the changed logic.

Reply:
```
async fn wait_until_ready<H: Hasher>(
    round: Round,
    creator: &mut Creator<H>,
    create_lag: &DelaySchedule,
    mut ignore_delay: bool,
    incoming_parents: &mut Receiver<Unit<H>>,
    mut exit: &mut oneshot::Receiver<()>,
) -> Result<(PreUnit<H>, Vec<H::Hash>), ()> {
    let mut delay = Delay::new(create_lag(round.into())).fuse();
    let mut result = if ignore_delay {
        creator.create_unit(round)
    } else {
        Err(())
    };
    while !ignore_delay || result.is_err() {
        futures::select! {
           unit = incoming_parents.next() => match unit {
                Some(unit) => {
                    creator.add_unit(&unit);
                    if ignore_delay {
                        result = creator.create_unit(round);
                    }
                },
                None => {
                    info!(target: "AlephBFT-creator", "Incoming parent channel closed, exiting.");
                    return Err(());
                }
            },
            _ = &mut delay => {
                if ignore_delay {
                    warn!(target: "AlephBFT-creator", "Delay passed despite us not waiting for it.");
                } else {
                    result = creator.create_unit(round);
                }
                ignore_delay = true;
                delay = Delay::new(Duration::from_secs(30 * 60)).fuse();
            },
            _ = exit => {
                info!(target: "AlephBFT-creator", "Received exit signal.");
                return Err(());
            },
        }
    }
    result
}

/// A process responsible for creating new units. It receives all the units added locally to the Dag
/// via the `incoming_parents` channel. It creates units according to an internal strategy respecting
/// always the following constraints: if round is equal to 0, U has no parents, otherwise for a unit U of round r > 0
/// - all U's parents are from round (r-1),
/// - all U's parents are created by different nodes,
/// - one of U's parents is the (r-1)-round unit by U's creator,
/// - U has > floor(2*N/3) parents.
/// - U will appear in the channel only if all U's parents appeared there before
/// The currently implemented strategy creates the unit U according to a delay schedule and when enough
/// candidates for parents are available for all the above constraints to be satisfied.
///
/// We refer to the documentation https://cardinal-cryptography.github.io/AlephBFT/internals.html
/// Section 5.1 for a discussion of this component.
pub async fn run<H: Hasher>(
    conf: Config,
    io: IO<H>,
    starting_round: oneshot::Receiver<Round>,
    mut exit: oneshot::Receiver<()>,
) {
    let Config {
        node_id,
        n_members,
        create_lag,
        max_round,
    } = conf;
    let mut creator = Creator::new(node_id, n_members);
    let IO {
        mut incoming_parents,
        outgoing_units,
    } = io;
    let starting_round = match starting_round.await {
        Ok(round) => round,
        Err(e) => {
            error!(target: "AlephBFT-creator", "Starting round not provided: {}", e);
            return;
        }
    };
    debug!(target: "AlephBFT-creator", "Creator starting from round {}", starting_round);
    for round in starting_round..max_round {
        // Skip waiting if we are way behind.
        let ignore_delay = creator.current_round() > round + 2 || round == 0;
        let new_unit = wait_until_ready(
            round,
            &mut creator,
            &create_lag,
            ignore_delay,
            &mut incoming_parents,
            &mut exit,
        )
        .await;
        let (unit, parent_hashes) = match new_unit {
            Err(_) => return,
            Ok(u) => u,
        };
        if let Err(e) =
            outgoing_units.unbounded_send(NotificationOut::CreatedPreUnit(unit, parent_hashes))
        {
            warn!(target: "AlephBFT-creator", "Notification send error: {}. Exiting.", e);
            return;
        }
    }
    warn!(target: "AlephBFT-creator", "Maximum round reached. Not creating another unit.");
}
```

Reply:
I really don't like how this mixes creation with waiting, especially with `create_unit` now being called in a couple places making it somewhat harder to reason about. :/

Reply:
After further discussions rewrote more or less like this, PTAL.